sdk/python/symbolchain/TypeParserBuilder.py

```diff
@@ -0,0 +1,98 @@
+import functools
+import importlib
+import operator
+
+from .TransactionDescriptorProcessor import TransactionDescriptorProcessor
+
+
+def _name_to_enum_value(mapping, enum_type, enum_value_name):
+	if enum_value_name not in mapping:
+		raise RuntimeError(f'unknown value {enum_value_name} for type {enum_type}')
+	return mapping[enum_value_name]
+
+
+def build_type_hints_map(transaction):
+	type_hints = {}
+	for key, hint in transaction.TYPE_HINTS.items():
+		rule_name = None
+		if hint.startswith('array['):
+			rule_name = hint
+		elif hint.startswith('enum:'):
+			rule_name = hint[len('enum:'):]
+		elif hint.startswith('pod:'):
+			rule_name = hint[len('pod:'):]
+		elif hint.startswith('struct:'):
+			rule_name = hint
+
+		if rule_name:
+			type_hints[key] = rule_name
+
+	return type_hints
+
+
+class TypeParserBuilder:
+	def __init__(self, name, type_converter, base_type_parsing_rules=None):
+		self.module_name = f'symbolchain.{name}'
+		self.type_converter = type_converter
+		self.base_type_parsing_rules = base_type_parsing_rules
+
+	def get_catbuffer_module(self):
+		return importlib.import_module(self.module_name)
+
+	def get_catbuffer_class(self, name):
+		return getattr(self.get_catbuffer_module(), name)
+
+	def create_flags_parser(self, name):
+		cls = self.get_catbuffer_class(name)
+		string_to_enum = dict(map(lambda key: (key.name.lower(), key), cls))
+
+		def parser(flags):
+			if isinstance(flags, str):
+				enum_array = list(map(lambda flag_name: _name_to_enum_value(string_to_enum, name, flag_name), flags.split(' ')))
+				return functools.reduce(operator.or_, enum_array)
+
+			if isinstance(flags, int):
+				return cls(flags)
+
+			return flags
+
+		return parser
+
+	def create_enum_parser(self, name):
+		cls = self.get_catbuffer_class(name)
+		string_to_enum = dict(map(lambda key: (key.name.lower(), key), cls))
+
+		def parser(enum_value):
+			if isinstance(enum_value, str):
+				return _name_to_enum_value(string_to_enum, name, enum_value)
+
+			if isinstance(enum_value, int):
+				return cls(enum_value)
+
+			return enum_value
+
+		return parser
+
+	def create_object_array_parser(self, name, existing_rules):
+		element_rule = existing_rules[name]
+
+		def parser(values):
+			return list(map(element_rule, values))
+
+		return parser
+
+	def create_struct_parser(self, struct_name, extend_type_parsing_rules):
```

Comment:
extend_type_parsing_rules => type_parsing_rules ?

Reply:
that is a function being passed here 😬